pkg/controller/ippool/pool_controller.go

```diff
@@ -0,0 +1,327 @@
+		if apierrors.IsNotFound(err) {
+			reqLogger.Info("Installation config not found")
+			r.status.OnCRNotFound()
+			return reconcile.Result{}, nil
+		}
+		reqLogger.Error(err, "An error occurred when querying the Installation resource")
+		return reconcile.Result{}, err
+	}
+
+	// Mark CR found so we can report converter problems via tigerastatus
+	r.status.OnCRFound()
+
+	// Get the APIServer. If healthy, we'll use it for managing pools.
+	// Otherwise, we'll use the CRD API for bootstrapping the cluster until the API server is available.
+	apiserver, _, err := utils.GetAPIServer(ctx, r.client)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
+	apiAvailable := apiserver != nil && apiserver.Status.State == v1.TigeraStatusReady
+
+	// SetMetaData in the TigeraStatus such as observedGenerations.
+	defer r.status.SetMetaData(&instance.ObjectMeta)
+
+	// Get all IP pools in the cluster.
+	currentPools := &crdv1.IPPoolList{}
+	err = r.client.List(ctx, currentPools)
+	if err != nil && !errors.IsNotFound(err) {
+		r.status.SetDegraded(operator.ResourceReadError, "error querying IP pools", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Create a lookup map of pools owned by this controller for easy access.
+	// This controller will ignore any IP pools that it itself did not create.
+	ourPools := map[string]crdv1.IPPool{}
+	for _, p := range currentPools.Items {
+		// TODO: Check if owned by this installation resource / controller.
+		ourPools[p.Spec.CIDR] = p
+	}
+
+	// For each pool that is desired, but doesn't exist, create it.
+	toCreate := []client.Object{}
+	for _, p := range instance.Spec.CalicoNetwork.IPPools {
+		if pool, ok := ourPools[p.CIDR]; !ok || !reflect.DeepEqual(pool, p) {
+			// Create a new pool, or update an existing one.
+			res := p.ToCRD()
+			if apiAvailable {
+				toCreate = append(toCreate, crdToV3(res))
+			} else {
+				toCreate = append(toCreate, res)
+			}
+		}
+	}
+
+	// Check existing pools owned by this controller that are no longer in
+	// the Installation resource.
+	toDelete := []client.Object{}
+	for cidr, pool := range ourPools {
+		reqLogger.WithValues("cidr", cidr).Info("Checking if pool is still valid")
+		found := false
+		for _, p := range instance.Spec.CalicoNetwork.IPPools {
+			if p.CIDR == cidr {
+				found = true
+				break
+			}
+		}
+		if !found {
+			reqLogger.WithValues("cidr", cidr).Info("Pool needs to be deleted")
+			if apiAvailable {
+				// No match. Needs delete. We only ever send deletes via the API server,
+				// since deletion requires rather complex logic. If the API server isn't available,
+				// we'll instead just mark the pool as disabled temporarily.
+				toDelete = append(toDelete, crdToV3(&pool))
+			} else {
+				// API server is not available. Just mark the pool as disabled so that new allocations
+				// don't come from this pool. We'll delete it once the API server is available.
+				pool.Spec.Disabled = true
```

Comment:
Similar to my comment about creating IPPools, would we really only need this during initial installation, where maybe the IPPools we configured incorrectly and the apiserver couldn't be brought up?

Reply:
It should be rare. I expect this will only be hit in cases where the apiserver is either accidentally not running, or because of some networking issue related to IP pools preventing the apiserver from working properly. In either case, we want to enable users to turn off IP pools in order to fix the system. Disabling the pool is a good compromise because it should effectively cordon the pool, but will wait for the apiserver to be running without finalizing the deletion (which is a complicated process that we don't want to replicate here)